### ISSUE TITLE
Unit type in Protelis

### ISSUE THREAD

**Issue by DanySK** [Protelis/Protelis#275]
Currently, Protelis treats void-returning methods the same way as null-returning methods, due to the way reflection is used internally. If the reflection mechanisms allow us to discriminate between void-returning methods and non non void-returning methods (and I believe they do), we could introduce unit-type in Protelis. This type would allow using calls to void methods anywhere in Protelis code.
Typical use case (by @jschewebbn):
```java
public void foo() { }
```
```python
def unitReturningFunction() {
  dostuff()
  foo()
}
```
currently, this would fail at runtime, complaining about null being returned.
Usual workaround (pre-protelis 13):
```python
def unitReturningFunction() {
  dostuff()
  foo()
  0
}
```
this is a kludge, though, and Protelis 13 still complains, as it's strictly rejecting null values. So the new "fix" would be:
```python
def unitReturningFunction() {
  dostuff()
  ofNullable(foo())
}
```
Which would systematically return an empty `Option`. This is conceptually better, but:

1. it's a lot of ceremony
2. `Option` will never be present, so it's actually abused for something else (namely, [unit type](https://en.wikipedia.org/wiki/Unit_type))

If @jakebeal and @mviroli agree, I'd propose to introduce Unit, and to do so directly in the interpreter. No parser changes required, as we got no static type checking anyway. This change should be almost entirely backwards compatible, the only difference would be that calls currently wrapped into an Option that will always be empty will be instead populated with a present Unit. But I believe the amount of code using Option is currently close to zero.

@jakebeal please assign a priority and milestone to this issue so that I can understand when to tackle it.

**Comment by jakebeal** [comment]
This looks like a good way to deal with the errors being introduced by the increased strictness in handling void in 13.x

**Comment by jschewebbn** [comment]
Would this cause the following Protelis code to be able to execute without a runtime null pointer?
```
def unitReturningFunction() {
  dostuff()
  foo()
}
```
Where `dostuff` is a Java function with a void return value?

**Comment by DanySK** [comment]
yes if both `dostuff` and `foo` return either a non-`null` value or are `void`. `null` values remain external and not allowed in Protelis.

**Comment by jschewebbn** [comment]
OK, that would make the upgrade to Protelis 13+ much cleaner. I would be able to stop adding all of these extra zero return values from methods.

**Comment by DanySK** [comment]
OK let's prioritize a 13.1.0 release with this feature kicked in then
